### DISCUSSION TITLE
MaxListenersExceededWarning - Angular 15

### DISCUSSION THREAD

**Discussion by i-am-the-arm** [mswjs/msw#1688] [Q&A]
Hello, I wanted to get some additional information on how this warning is generated:

`MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 response:mocked listeners added. Use emitter.setMaxListeners() to increase limit`

I have a page that makes several independent api requests on load and seems to throw this warning when this happens. When I only have 1 or 2 requests, this warning goes away. Can anyone explain how `msw` handles or adds listeners?

- Is this truly an `msw` warning or does it come from the underlying framework?
- Does `msw` create new listeners for every request that's made? (i.e., if I have 15 api requests, should I expect to have 15 listeners; even if the same endpoint is called more than once?)

Ultimately, I just want to ensure there's not a concern of memory leaks but wasn't sure where to look to validate.

Thanks in advance!

**Comment by dgilleland** [comment]
Hi @i-am-the-arm ,

## Diagnosis

I know your question was quite a while ago. I just ran into this same warning myself for the first time when building out a suite of unit tests. I'm new to msw, but my story might help you (or others finding this thread).

To better understand what's going on, check the details of the warning to see if there is a specific listener that is the culprit.  In my case, it was all the `server.events.on('request:start', ...)` listeners I was creating:

```
  emitter: _Emitter {
    events: Map(1) { 'request:start' => [Array] },
    maxListeners: 10,
    hasWarnedAboutPotentialMemoryLeak: true
  },
  type: 'request:start',
  count: 11
```

I know that in the docs for msw it states "You are unlikely to use this API directly while developing or testing with MSW". I don't know how often people actually need to peek into the server events to see what's going on, but I suspect that it should be infrequent. (My situation was unique...)

## Fix

**The solution** for *my* situation was to clear out the listener after each test:

```js
    afterEach(() => {
        server.events.removeAllListeners();
        server.resetHandlers();
    });
```

## RnD

Here are some possible articles to deepen your understanding of the `MaxListenersExceededWarning`:

- [🚀 MaxListenersExceededWarning: Ultimate Guide to Fixing EventEmitter Memory Leak Issues](https://lxadm.com/maxlistenersexceededwarning-possible-eventemitter-memory-leak-detected/)
- [[Solved]: MaxListenersExceededWarning — Understanding and Resolving the EventEmitter Memory Leak](https://medium.com/@zahidbashirkhan/solved-maxlistenersexceededwarning-understanding-and-resolving-the-eventemitter-memory-leak-93df6ff4b5d4)
